packages/toolkit/src/query/core/buildMiddleware/cacheCollection.ts

```diff
@@ -66,9 +67,18 @@ export const buildCacheCollectionHandler: InternalHandlerBuilder = ({
     mwApi,
     internalState
   ) => {
-    if (unsubscribeQueryResult.match(action)) {
+    if (unsubscribeQueryResult.match(action) ||
+      queryThunk.fulfilled.match(action) ||
+      queryThunk.rejected.match(action)
+    ) {
       const state = mwApi.getState()[reducerPath]
-      const { queryCacheKey } = action.payload
+      let queryCacheKey: QueryCacheKey
+
```

Comment:
```suggestion
  const canTriggerUnsubscribe = isAnyOf(
    unsubscribeQueryResult.match,
    queryThunk.fulfilled,
    queryThunk.rejected
  )
  const handler: ApiMiddlewareInternalHandler = (action, mwApi) => {
    if (canTriggerUnsubscribe(action)) {
      const state = mwApi.getState()[reducerPath]
      const { queryCacheKey } = unsubscribeQueryResult.match(action)
        ? action.payload
        : action.meta.arg
```

Let's get this down in size even a little bit further.

Reply:
Amazing! This is what I was trying to wrack my brain to come up with originally, but forgot to come back to it.
Newest commit should reflect this change.